contracts/EtomicSwapV2.sol

```diff
@@ -0,0 +1,483 @@
+                takerSecretHash,
+                ripemd160(abi.encodePacked(sha256(abi.encodePacked(makerSecret)))),
+                tokenAddress
+            )
+        );
+        require(paymentHash == makerPayments[id].paymentHash, "Invalid paymentHash");
+
+        makerPayments[id].state = MakerPaymentState.TakerSpent;
+
+        if (tokenAddress == address(0)) {
+            payable(msg.sender).transfer(amount);
+        } else {
+            IERC20 token = IERC20(tokenAddress);
+            require(
+                token.transfer(msg.sender, amount), "ERC20 transfer failed: Contract may lack balance or token transfer was rejected"
+            );
+        }
+    }
+
+    function refundMakerPaymentTimelock(
+        bytes32 id,
+        uint256 amount,
+        address taker,
+        bytes20 takerSecretHash,
+        bytes20 makerSecretHash,
+        address tokenAddress
+    ) external {
+        require(
+            makerPayments[id].state == MakerPaymentState.PaymentSent,
+            "Invalid payment state. Must be PaymentSent"
+        );
+
+        bytes20 paymentHash = ripemd160(
+            abi.encodePacked(
+                amount,
+                taker,
+                msg.sender,
+                takerSecretHash,
+                makerSecretHash,
+                tokenAddress
+            )
+        );
+
+        require(
+            paymentHash == makerPayments[id].paymentHash,
+            "Invalid paymentHash"
+        );
+
+        require(
+            block.timestamp >= makerPayments[id].paymentLockTime,
+            "Current timestamp didn't exceed payment refund lock time"
+        );
+
+        makerPayments[id].state = MakerPaymentState.MakerRefunded;
+
+        emit MakerPaymentRefundedTimelock(id);
+
+        if (tokenAddress == address(0)) {
+            payable(msg.sender).transfer(amount);
+        } else {
+            IERC20 token = IERC20(tokenAddress);
+            require(token.transfer(msg.sender, amount));
+        }
+    }
+
+    function refundMakerPaymentSecret(
+        bytes32 id,
+        uint256 amount,
+        address taker,
+        bytes32 takerSecret,
+        bytes20 makerSecretHash,
+        address tokenAddress
+    ) external {
+        require(
+            makerPayments[id].state == MakerPaymentState.PaymentSent,
+            "Invalid payment state. Must be PaymentSent"
+        );
+
+        bytes20 paymentHash = ripemd160(
+            abi.encodePacked(
+                amount,
+                taker,
+                msg.sender,
+                ripemd160(abi.encodePacked(sha256(abi.encodePacked(takerSecret)))),
+                makerSecretHash,
+                tokenAddress
+            )
+        );
+
+        require(
+            paymentHash == makerPayments[id].paymentHash,
+            "Invalid paymentHash"
+        );
+
+        makerPayments[id].state = MakerPaymentState.MakerRefunded;
+
+        emit MakerPaymentRefundedSecret(id, takerSecret);
```

Comment:
Do we need to log the takerSecret? I don't think it will be needed after maker payment is refunded.

Reply:
Good note, it will likely slightly reduce gas used, thanks 🙂